packages/flutter/lib/src/widgets/semantics_debugger.dart

```diff
@@ -262,79 +300,51 @@ class _SemanticsDebuggerListener implements mojom.SemanticsListener {
-    _listeners.add(callback);
-  }
-  void removeListener(VoidCallback callback) {
-    _listeners.remove(callback);
-  }
-
-  Map<int, _SemanticsDebuggerEntry> nodes = <int, _SemanticsDebuggerEntry>{};
+  _SemanticsDebuggerEntry get rootNode => _nodes[0];
+  final Map<int, _SemanticsDebuggerEntry> _nodes = <int, _SemanticsDebuggerEntry>{};
 
   _SemanticsDebuggerEntry _updateNode(mojom.SemanticsNode node) {
-    _SemanticsDebuggerEntry entry = nodes.putIfAbsent(node.id, () => new _SemanticsDebuggerEntry(node.id));
-    if (node.flags != null) {
-      entry.canBeTapped = node.flags.canBeTapped;
-      entry.canBeLongPressed = node.flags.canBeLongPressed;
-      entry.canBeScrolledHorizontally = node.flags.canBeScrolledHorizontally;
-      entry.canBeScrolledVertically = node.flags.canBeScrolledVertically;
-      entry.hasCheckedState = node.flags.hasCheckedState;
-      entry.isChecked = node.flags.isChecked;
-    }
-    if (node.strings != null) {
-      assert(node.strings.label != null);
-      entry.label = node.strings.label;
-    } else {
-      assert(entry.label != null);
-    }
-    if (node.geometry != null) {
-      if (node.geometry.transform != null) {
-        assert(node.geometry.transform.length == 16);
-        // TODO(ianh): Replace this with a cleaner call once
-        //  https://github.com/google/vector_math.dart/issues/159
-        // is fixed.
-        List<double> array = node.geometry.transform;
-        entry.transform = new Matrix4(
-          array[0],  array[1],  array[2],  array[3],
-          array[4],  array[5],  array[6],  array[7],
-          array[8],  array[9],  array[10], array[11],
-          array[12], array[13], array[14], array[15]
-        );
-      } else {
-        entry.transform = null;
-      }
-      entry.rect = new Rect.fromLTWH(node.geometry.left, node.geometry.top, node.geometry.width, node.geometry.height);
-    }
-    entry.updateMessage();
+    final int id = node.id;
+    _SemanticsDebuggerEntry entry = _nodes.putIfAbsent(id, () => new _SemanticsDebuggerEntry(id));
+    entry.updateWith(node);
     if (node.children != null) {
-      Set<_SemanticsDebuggerEntry> oldChildren = new Set<_SemanticsDebuggerEntry>.from(entry.children ?? const <_SemanticsDebuggerEntry>[]);
-      entry.children?.clear();
-      entry.children ??= new List<_SemanticsDebuggerEntry>();
+      if (entry.children != null)
+        entry.children.clear();
+      else
+        entry.children = new List<_SemanticsDebuggerEntry>();
       for (mojom.SemanticsNode child in node.children)
         entry.children.add(_updateNode(child));
-      Set<_SemanticsDebuggerEntry> newChildren = new Set<_SemanticsDebuggerEntry>.from(entry.children);
-      Set<_SemanticsDebuggerEntry> removedChildren = oldChildren.difference(newChildren);
-      for (_SemanticsDebuggerEntry oldChild in removedChildren)
-        nodes.remove(oldChild.id);
     }
     return entry;
   }
 
+  void _removeDetachedNodes() {
```

Comment:
maybe add a TODO to the effect that we should ideally find a way to avoid walking the entire tree to do this